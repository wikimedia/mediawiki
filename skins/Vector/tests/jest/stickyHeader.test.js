const mustache = require( 'mustache' );
const fs = require( 'fs' );
const stickyHeaderTemplate = fs.readFileSync( 'includes/templates/StickyHeader.mustache', 'utf8' );
const buttonTemplate = fs.readFileSync( 'includes/templates/Button.mustache', 'utf8' );
const iconTemplate = fs.readFileSync( 'includes/templates/Icon.mustache', 'utf8' );
const sticky = require( '../../resources/skins.vector.js/stickyHeader.js' );
const { userLinksHTML, dropdownPartials } = require( './userLinksData.js' );

const defaultButtonsTemplateData = [ {
	href: '#',
	id: 'ca-talk-sticky-header',
	icon: 'speechBubbles',
	class: 'mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon',
	'array-attributes': [ {
		key: 'data-event-name',
		value: 'talk-sticky-header'
	}, {
		key: 'tabindex',
		value: '-1'
	} ]
}, {
	href: '#',
	id: 'ca-history-sticky-header',
	icon: 'history',
	class: 'mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon',
	'array-attributes': [ {
		key: 'data-event-name',
		value: 'history-sticky-header'
	}, {
		key: 'tabindex',
		value: '-1'
	} ]
}, {
	href: '#',
	id: 'ca-watchstar-sticky-header',
	icon: 'star',
	class: 'mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon mw-watchlink',
	'array-attributes': [ {
		key: 'data-event-name',
		value: 'watch-sticky-header'
	}, {
		key: 'tabindex',
		value: '-1'
	} ]
} ];

const editButtonsTemplateData = [ {
	href: '#',
	id: 'ca-ve-edit-sticky-header',
	icon: 'edit',
	class: 'mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon',
	'array-attributes': [ {
		key: 'data-event-name',
		value: 've-edit-sticky-header'
	}, {
		key: 'tabindex',
		value: '-1'
	} ]
}, {
	href: '#',
	id: 'ca-edit-sticky-header',
	icon: 'wikiText',
	class: 'mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon',
	'array-attributes': [ {
		key: 'data-event-name',
		value: 'wikitext-edit-sticky-header'
	}, {
		key: 'tabindex',
		value: '-1'
	} ]
}, {
	href: '#',
	id: 'ca-viewsource-sticky-header',
	icon: 'star',
	class: 'mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon',
	'array-attributes': [ {
		key: 'data-event-name',
		value: 'editLock'
	}, {
		key: 'tabindex',
		value: '-1'
	} ]
} ];

const templateData = {
	'data-toc': {
		'array-sections': []
	},
	'data-sticky-header-toc-dropdown': {
		id: 'vector-sticky-header-toc',
		class: 'mw-portlet mw-portlet-sticky-header-toc vector-sticky-header-toc',
		'html-items': '',
		'html-vector-menu-checkbox-attributes': 'tabindex="-1"',
		'html-vector-menu-heading-attributes': 'tabindex="-1"',
		'heading-class': 'mw-ui-button mw-ui-quiet mw-ui-icon mw-ui-icon-element mw-ui-icon-wikimedia-listBullet'
	},
	'array-buttons': [ {
		label: '0 languages',
		id: 'p-lang-btn-sticky-header',
		class: 'mw-ui-button mw-ui-quiet mw-interlanguage-selector',
		'array-attributes': [ {
			key: 'data-event-name',
			value: 'ui.dropdown-p-lang-btn-sticky-header'
		}, {
			key: 'tabindex',
			value: '-1'
		} ]
	} ],
	'data-button-start': {
		label: 'search',
		icon: 'search',
		class: 'mw-ui-button mw-ui-quiet mw-ui-icon mw-ui-icon-element vector-sticky-header-search-toggle',
		'array-attributes': [ {
			key: 'data-event-name',
			value: 'ui.vector-sticky-search-form.icon'
		}, {
			key: 'tabindex',
			value: '-1'
		} ]
	},
	'data-search': {},
	'array-icon-buttons': defaultButtonsTemplateData.concat( editButtonsTemplateData )
};

const renderedHTML = mustache.render(
	stickyHeaderTemplate, templateData, Object.assign( {}, dropdownPartials, {
		Button: buttonTemplate,
		Icon: iconTemplate,
		SearchBox: '<div> </div>' // ignore SearchBox for this test
	} ) );

beforeEach( () => {
	document.body.innerHTML = renderedHTML;
} );

test( 'Sticky header renders', () => {
	expect( document.body.innerHTML ).toMatchSnapshot();
} );

describe( 'sticky header', () => {
	test( 'prepareUserLinksDropdown removes gadgets from dropdown', async () => {
		const menu = document.createElement( 'div' );
		menu.innerHTML = userLinksHTML;
		const userLinksDropdown = /** @type {Element} */ ( menu.querySelector( '#' + sticky.USER_LINKS_DROPDOWN_ID ) );
		const newMenu = sticky.prepareUserLinksDropdown( userLinksDropdown );
		// check classes have been updated and removed.
		expect( userLinksDropdown.querySelectorAll( '.user-links-collapsible-item' ).length > 0 ).toBeTruthy();
		expect( userLinksDropdown.querySelectorAll( '.mw-list-item-js' ).length > 0 ).toBeTruthy();
		expect( newMenu.querySelectorAll( '.user-links-collapsible-item' ).length ).toBe( 0 );
		expect( newMenu.querySelectorAll( '.mw-list-item-js' ).length ).toBe( 0 );
	} );
} );
