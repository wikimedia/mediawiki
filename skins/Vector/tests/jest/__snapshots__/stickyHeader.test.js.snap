// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sticky header renders 1`] = `
"<div id=\\"vector-sticky-header\\" class=\\"vector-sticky-header\\">
	<div class=\\"vector-sticky-header-start\\">
		<div class=\\"vector-sticky-header-icon-start mw-ui-icon-flush-left mw-ui-icon-flush-right\\" aria-hidden=\\"true\\">

<button class=\\"mw-ui-button mw-ui-quiet mw-ui-icon mw-ui-icon-element vector-sticky-header-search-toggle\\" data-event-name=\\"ui.vector-sticky-search-form.icon\\" tabindex=\\"-1\\"><span class=\\"mw-ui-icon mw-ui-icon-search mw-ui-icon-wikimedia-search\\"></span>

<span>search</span>

</button>
		</div>
<div> </div>		<div class=\\"vector-sticky-header-context-bar\\">
			<nav role=\\"navigation\\" aria-label=\\"\\" class=\\"vector-toc-landmark\\">
<div id=\\"vector-sticky-header-toc\\" class=\\"vector-menu vector-dropdown vector-menu-dropdown mw-portlet mw-portlet-sticky-header-toc vector-sticky-header-toc\\">
	<input type=\\"checkbox\\" id=\\"vector-sticky-header-toc-checkbox\\" role=\\"button\\" aria-haspopup=\\"true\\" data-event-name=\\"ui.dropdown-vector-sticky-header-toc\\" class=\\"vector-menu-checkbox\\" aria-label=\\"\\" tabindex=\\"-1\\">
	<label id=\\"vector-sticky-header-toc-label\\" for=\\"vector-sticky-header-toc-checkbox\\" class=\\"vector-menu-heading mw-ui-button mw-ui-quiet mw-ui-icon mw-ui-icon-element mw-ui-icon-wikimedia-listBullet\\" aria-hidden=\\"true\\" tabindex=\\"-1\\">
		<span class=\\"vector-menu-heading-label\\"></span>
	</label>
	<div class=\\"vector-menu-content vector-dropdown-content\\">

	</div>
</div>			</nav>
			<div class=\\"vector-sticky-header-context-bar-primary\\" aria-hidden=\\"true\\"></div>
		</div>
	</div>
	<div class=\\"vector-sticky-header-end\\" aria-hidden=\\"true\\">
		<div class=\\"vector-sticky-header-icons\\">
<a href=\\"#\\" id=\\"ca-talk-sticky-header\\" class=\\"mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon\\" data-event-name=\\"talk-sticky-header\\" tabindex=\\"-1\\"><span class=\\"mw-ui-icon mw-ui-icon-speechBubbles mw-ui-icon-wikimedia-speechBubbles\\"></span>

<span></span>
</a>

<a href=\\"#\\" id=\\"ca-history-sticky-header\\" class=\\"mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon\\" data-event-name=\\"history-sticky-header\\" tabindex=\\"-1\\"><span class=\\"mw-ui-icon mw-ui-icon-history mw-ui-icon-wikimedia-history\\"></span>

<span></span>
</a>

<a href=\\"#\\" id=\\"ca-watchstar-sticky-header\\" class=\\"mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon mw-watchlink\\" data-event-name=\\"watch-sticky-header\\" tabindex=\\"-1\\"><span class=\\"mw-ui-icon mw-ui-icon-star mw-ui-icon-wikimedia-star\\"></span>

<span></span>
</a>

<a href=\\"#\\" id=\\"ca-ve-edit-sticky-header\\" class=\\"mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon\\" data-event-name=\\"ve-edit-sticky-header\\" tabindex=\\"-1\\"><span class=\\"mw-ui-icon mw-ui-icon-edit mw-ui-icon-wikimedia-edit\\"></span>

<span></span>
</a>

<a href=\\"#\\" id=\\"ca-edit-sticky-header\\" class=\\"mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon\\" data-event-name=\\"wikitext-edit-sticky-header\\" tabindex=\\"-1\\"><span class=\\"mw-ui-icon mw-ui-icon-wikiText mw-ui-icon-wikimedia-wikiText\\"></span>

<span></span>
</a>

<a href=\\"#\\" id=\\"ca-viewsource-sticky-header\\" class=\\"mw-ui-button mw-ui-quiet mw-ui-icon-element sticky-header-icon\\" data-event-name=\\"editLock\\" tabindex=\\"-1\\"><span class=\\"mw-ui-icon mw-ui-icon-star mw-ui-icon-wikimedia-star\\"></span>

<span></span>
</a>

		</div>
		<div class=\\"vector-sticky-header-buttons\\">

<button id=\\"p-lang-btn-sticky-header\\" class=\\"mw-ui-button mw-ui-quiet mw-interlanguage-selector\\" data-event-name=\\"ui.dropdown-p-lang-btn-sticky-header\\" tabindex=\\"-1\\"><span>0 languages</span>

</button>
		</div>
		<div class=\\"vector-sticky-header-icon-end\\">
			<div class=\\"vector-user-links\\">
			</div>
		</div>
	</div>
</div>
"
`;
